### ISSUE TITLE
Handshake S0 crash on AWS MediaLive RTMP transmission

### ISSUE THREAD

**Issue by andrew3213221** [pedroSG94/rtmp-rtsp-stream-client-java#932]
We use AWS MediaLive https://aws.amazon.com/medialive/ for RTMP live transmission. 
Everything is working as expected, but we get the following error (reproducible) if we do the following: 
* prepare encoders 🆗 
* connect to RTMP endpoint 🆗 
* send media data (audio and video) 🆗 
* stop the connection 🆗 
* stop encoders 🆗 
* prepare encoders (because we want to connect again) 🆗 
* connection to RTMP endpoint 🛑  CRASH

Sidenote: the same works with YouTube RTMP connection without any problem. 

Used `rtmp-rtsp-stream-client-java` version: **2.0.7**

Log of our Android (Android 7.1.1): 
```log
I/VideoEncoder: 2 encoders found
I/VideoEncoder: Encoder OMX.Nvidia.h264.encoder
I/VideoEncoder: Color supported: 2135033992
    Color supported: 19
    Color supported: 21
    Color supported: 2130708361
    Encoder selected OMX.Nvidia.h264.encoder
I/OMXClient: MuxOMX ctor
I/VideoEncoder: Prepare video info: SURFACE, 1280x720
I/VideoEncoder: bitrate mode CBR not supported using default mode
E/ACodec: [OMX.Nvidia.h264.encoder] storeMetaDataInBuffers (output) failed w/ err -2147483648
W/ACodec: do not know color format 0x7f000789 = 2130708361
I/ACodec: setupAVCEncoderParameters with [profile: Baseline] [level: Level5]
I/ACodec: [OMX.Nvidia.h264.encoder] cannot encode color aspects. Ignoring.
    [OMX.Nvidia.h264.encoder] cannot encode HDR static metadata. Ignoring.
    setupVideoEncoder succeeded
W/ACodec: do not know color format 0x7f000789 = 2130708361
I/VideoEncoder: prepared
E/SurfaceManager: GL already released
I/SurfaceManager: GL initialized
E/SurfaceManager: GL already released
I/SurfaceManager: GL initialized
I/SurfaceManager: GL released
E/SurfaceManager: GL already released
I/SurfaceManager: GL initialized
I/SurfaceManager: GL initialized
I/MicrophoneManager: Microphone created, 32000hz, Stereo
I/AudioEncoder: 1 encoders found
    Encoder OMX.google.aac.encoder
    Encoder selected OMX.google.aac.encoder
I/OMXClient: MuxOMX ctor
I/AudioEncoder: prepared
V/Camera: No recordingCaptureSession to close
V/Camera: open successful 
I/Handshake: writing C0
    C0 write successful
    writing C1
    writing time 1630676498 to c1
    writing zero to c1
    writing random to c1
I/Handshake: C1 write successful
I/Handshake: reading S0
E/RtmpClient: connection error
    java.io.IOException: Handshake error, unexpected -1 S0 received
        at com.pedro.rtmp.rtmp.Handshake.readS0(Handshake.kt:126)
        at com.pedro.rtmp.rtmp.Handshake.sendHandshake(Handshake.kt:64)
        at com.pedro.rtmp.rtmp.RtmpClient.establishConnection(RtmpClient.kt:218)
        at com.pedro.rtmp.rtmp.RtmpClient.connect$lambda-0(RtmpClient.kt:145)
        at com.pedro.rtmp.rtmp.RtmpClient.$r8$lambda$2Ex0pWOcGVHPFRJWQRhtgifpZyQ(RtmpClient.kt)
        at com.pedro.rtmp.rtmp.RtmpClient$$ExternalSyntheticLambda0.run(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
        at java.lang.Thread.run(Thread.java:761)
E/RTMP: Rtmp connection failed Error configure stream, Handshake error, unexpected -1 S0 received
```

What we do wrong? Any suggestion?
Happy to get your feedback!
Best, Andrew

**Comment by pedroSG94** [comment]
Can you send me a guide to configure stream exactly like you and reproduce your error?
Also, if you want, you can share me an endpoint to reproduce the problem. You can send it to my email to keep it private:
pedrossgg1994@gmail.com

**Comment by andrew3213221** [comment]
I sent you to your email

**Comment by pedroSG94** [comment]
Hello,

I did a try with my app example and all is working fine.

I received a custom code used in my email. Can you try compile my app example and check if you can reproduce this error?
Are you sure that RtmpClient.disconnect was called corerctly? can you share a full logcat?

**Comment by andrew3213221** [comment]
I will try at Monday

**Comment by andrew3213221** [comment]
With example app it works.
I am using custom code cause i can't find the way how i can add your own Surface to the CaptureSession with RtmpCamera1 or RtmpCamera2.
Could you help me with is?

**Comment by andrew3213221** [comment]
For close the stream i am using RtmpClient.disconnect() is it correct?

**Comment by andrew3213221** [comment]
log on close 

`E/RtmpClient: disconnect error
    java.net.SocketException: Socket closed
        at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:120)
        at java.net.SocketOutputStream.write(SocketOutputStream.java:157)
        at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
        at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
        at com.pedro.rtmp.rtmp.CommandsManager.sendClose(CommandsManager.kt:201)
        at com.pedro.rtmp.rtmp.RtmpClient.disconnect$lambda-5(RtmpClient.kt:415)
        at com.pedro.rtmp.rtmp.RtmpClient.$r8$lambda$L6f1XZSd9tdbDvwE21RqUPtnFoA(RtmpClient.kt)
        at com.pedro.rtmp.rtmp.RtmpClient$$ExternalSyntheticLambda1.run(Unknown Source)`

**Comment by pedroSG94** [comment]
This is weird.
In CommandsManager.kt:201 close command is already sent but you had no time to flush because server close connection too fast.
This isn't a error so all is working as expected.
About your implementation try this.
- Copy Camera2Base.
- [In this line](https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtplibrary/src/main/java/com/pedro/rtplibrary/base/Camera2Base.java#L118) add your TextureView as a parameters.
- Now set this textureview to Camera2ApiManager like with glInterface (remember modify that method in Camera2ApiManager):
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtplibrary/src/main/java/com/pedro/rtplibrary/base/Camera2Base.java#L454
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtplibrary/src/main/java/com/pedro/rtplibrary/base/Camera2Base.java#L490
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/master/rtplibrary/src/main/java/com/pedro/rtplibrary/base/Camera2Base.java#L608
After that you can add your new textureview of Camera2ApiManager to render list:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/blob/226b1a56f0ac5dfecd09a1aa401046232c4d514f/encoder/src/main/java/com/pedro/encoder/input/video/Camera2ApiManager.java#L128

**Comment by andrew3213221** [comment]
I copy all required classes and i found following problem
1. RtmpSender write just audio video packet is missing log=> 
`I/RtmpSender: wrote Audio packet, size 270
I/RtmpSender: wrote Audio packet, size 266
I/RtmpSender: wrote Audio packet, size 279
I/RtmpSender: wrote Audio packet, size 270
I/RtmpSender: wrote Audio packet, size 279
I/RtmpSender: wrote Audio packet, size 280
I/RtmpSender: wrote Audio packet, size 263
I/RtmpSender: wrote Audio packet, size 276
I/RtmpSender: wrote Audio packet, size 300
I/RtmpSender: wrote Audio packet, size 275
I/RtmpSender: wrote Audio packet, size 316`
2. The same problem on close disconnect error log => 
`disconnect error
    java.net.SocketException: Socket closed
        at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:120)
        at java.net.SocketOutputStream.write(SocketOutputStream.java:157)
        at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
        at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
        at com.pedro.rtmp.rtmp.CommandsManager.sendClose(CommandsManager.kt:201)
        at com.pedro.rtmp.rtmp.RtmpClient.disconnect$lambda-5(RtmpClient.kt:415)
        at com.pedro.rtmp.rtmp.RtmpClient.$r8$lambda$L6f1XZSd9tdbDvwE21RqUPtnFoA(RtmpClient.kt)`

I opened example app **surface mode** where used rtmpCamera2 and got the same problems

**Comment by pedroSG94** [comment]
Hello,

As I said that error is not a problem itself. That means that flush of outputbuffer failed but close command was send correctly. If you can connect to the server after that, you can ignore it (Maybe I should change that error mensaje to avoid confuse users).

I think that the problem is not related with disconnect. The weird is that you can't connect using AWS only with your custom implementation for some reason.
Can you share me a full example of your code? Not only custom implementation, I need an activity example to compile and test it.

**Comment by pedroSG94** [comment]
I did a commit that will remove disconnect error. I don't think that your problem is related with that because your send close is properly send but you can try with that:
https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/commit/639ad549498ae434918969c862dddb2602a2c67a
Gradle version:
```gradle
implementation 'com.github.pedroSG94:rtmp-rtsp-stream-client-java:639ad54949'

```
If your problem is not resolved. Share me logcat mentioned in my last post.

**Comment by andrew3213221** [comment]
Hello thanks for the feedback,
What about Rtmp Sender wrote audio packet, video packet is missing.
Server doesn't see the livestream.
I tested on example app Surface mode and it doesn't work there.
Could you check it?

**Comment by pedroSG94** [comment]
Camera2 with SurfaceView or TextureView is marked as deprecated because I can't control orientation/aspect ratio. Also, use 2 surfaces is not surpported by all devices producing problems
Anyway, it is working for me. Do you have a logcat of Camera2ApiManager to know the error?
Is your preview working fine?

**Comment by Eittipat** [comment]
Hello,
This may relate to this issue.

At first, I have the same "Socket closed" error on version 2.0.7.
Then, I switched to commit "639ad54949" and got this problem instead

```
E/AndroidRuntime( 9630): java.lang.StackOverflowError: stack size 8192KB
E/AndroidRuntime( 9630): 	at java.lang.ref.FinalizerReference.<init>(FinalizerReference.java:48)
E/AndroidRuntime( 9630): 	at java.lang.ref.FinalizerReference.add(FinalizerReference.java:61)
E/AndroidRuntime( 9630): 	at java.util.concurrent.Executors.newSingleThreadExecutor(Executors.java:179)
E/AndroidRuntime( 9630): 	at com.pedro.rtmp.rtmp.RtmpClient.disconnect(RtmpClient.kt:405)
E/AndroidRuntime( 9630): 	at com.pedro.rtmp.rtmp.RtmpClient.disconnect(RtmpClient.kt:399)
E/AndroidRuntime( 9630): 	at com.whelksoft.camera_with_rtmp.RtmpCameraConnector.stopStreamRtp(RtmpCameraConnector.kt:359)
E/AndroidRuntime( 9630): 	at com.whelksoft.camera_with_rtmp.RtmpCameraConnector.stopStream(RtmpCameraConnector.kt:257)
E/AndroidRuntime( 9630): 	at com.whelksoft.camera_with_rtmp.Camera.onDisconnectRtmp(Camera.kt:742)
E/AndroidRuntime( 9630): 	at com.whelksoft.camera_with_rtmp.RtmpCameraConnector.onDisconnectRtmp(RtmpCameraConnector.kt:501)
E/AndroidRuntime( 9630): 	at com.pedro.rtmp.rtmp.RtmpClient.disconnect(RtmpClient.kt:430)
E/AndroidRuntime( 9630): 	at com.pedro.rtmp.rtmp.RtmpClient.disconnect(RtmpClient.kt:399)
....

```

P.S. I'm using my own customized RTMP server, but everything works fine with version 1.8.2

**Comment by andrew3213221** [comment]
Thank you, everything works with '639ad54949' no errors
